import "common/request.proto";

package specpaxos.fastpaxos.proto;

message RequestMessage {
    required specpaxos.Request req = 1;
}

message ReplyMessage {
    required uint64 view = 1;
    required uint64 opnum = 2;
    required bytes reply = 3;
    required uint64 clientreqid = 4;
}

message UnloggedRequestMessage {
    required specpaxos.UnloggedRequest req = 1;
}

message UnloggedReplyMessage {
    required bytes reply = 1;
}

message PrepareOKMessage {
    required uint64 view = 1;
    required uint64 opnum = 2;
    required uint32 replicaIdx = 3;
    required specpaxos.Request req = 4;
    required uint32 slowpath = 5;
}

message PrepareMessage {
    required uint64 view = 1;
    required uint64 opnum = 2;
    required specpaxos.Request req = 3;
}

message CommitMessage {
    required uint64 view = 1;
    required uint64 opnum = 2;
    required specpaxos.Request req = 3;
}

message RequestStateTransferMessage {
    required uint64 view = 1;
    required uint64 opnum = 2;
}

message StateTransferMessage {
    required uint64 view = 1;
    required uint64 opnum = 2;
    repeated specpaxos.MsgLogEntry entries = 3;
    required uint64 lastop = 4;
}

